notebooks/name-search-utility-demo.ipynb

```diff
@@ -0,0 +1,331 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 3. Import named segments\n",
+    "### 3a. Import the raw .json file\n",
+    "This file is a dictionary organized as:\n",
+    "```\n",
+    "{\"Name of River as Key Value\":[list,of,nwm,link,ids,with,that,name],\n",
+    " \"Another Name River\":[xxxx,yyyy,etc.]}\n",
+    "```\n",
+    "That structure minimizes duplication. However, it should be noted that it also means that every Sand Creek is listed under the same `\"Sand Creek\"` key and that care must be taken in interpreting the results."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "json_names_path = os.path.join(\n",
+    "    root, r\"test\", \"input\", \"json\", r\"nwm_reaches_conus_v21_wgnis_name.json\",\n",
+    ")\n",
+    "\n",
+    "with open(json_names_path, \"r\") as json_file:\n",
+    "    names = json.load(json_file)\n",
+    "    if names:\n",
+    "        print(f\"success!! -- imported {len(names)} GNIS names from the table.\")\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3b. Convert the names into a dataframe keyed on segment id"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ids_wnames = {}\n",
+    "for name, ids in names.items():\n",
+    "    for id in ids:\n",
+    "        ids_wnames[id] = name\n",
+    "\n",
+    "ids_wnames_df = pd.DataFrame.from_dict(ids_wnames, orient=\"index\", columns=[\"name\"])\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 4. Import the NWM segments \n",
+    "### 4a. Import the raw table (a.k.a. 'the Route/Link' file) as a dataframe\n",
+    "This is the same file used for driving the channel computations of the National Water Model. Other scripts in this repository will download it automatically for you in the background -- here, we're just a little more practical: if you don't have it, just do this: \n",
+    "```\n",
+    "export ROUTELINK=\"RouteLink_NHDPLUS.nc\"\n",
+    "wget -c https://www.nco.ncep.noaa.gov/pmb/codes/nwprod/nwm.v2.0.2/parm/domain/$ROUTELINK\n",
+    "nccopy -d1 -s $ROUTELINK ${ROUTELINK/\\.nc/_compressed.nc}\n",
+    "```\n",
+    "Then copy the resulting file into the `geo_file_path` given below."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "geo_file_path = os.path.join(root, \"test\", \"input\", \"geo\", \"Channels\",)\n",
+    "\n",
+    "full_nc_path = os.path.join(geo_file_path, \"RouteLink_NHDPLUS.nwm.v2.0.2.nc\",)\n",
+    "\n",
+    "key_col = \"link\"\n",
+    "\n",
+    "rows = None\n",
+    "rows = (xr.open_dataset(full_nc_path)).to_dataframe()\n",
+    "rows = rows.set_index([key_col])\n",
+    "if rows is not None:\n",
```

Comment:
If there was an error creating the dataframe, an exception would have been thrown before reaching this block.